cpp/convert_json_to_marc.cc

```diff
@@ -0,0 +1,493 @@
+#include <vector>
+#include "FileUtil.h"
+#include "IniFile.h"
+#include "JSON.h"
+#include "MARC.h"
+#include "RegexMatcher.h"
+#include "util.h"
+
+
+namespace {
+
+
+struct FieldDescriptor {
+    std::string name_;
+    std::string tag_, overflow_tag_;
+    char indicator1_, indicator2_;
+    bool repeat_field_;
+    std::vector<std::pair<char, std::string>> subfield_codes_to_json_tags_, subfield_codes_to_prefixes_,
+                                              subfield_codes_to_fixed_subfields_; // For mapping to variable fields
+    std::string json_tag_; // For mapping to control fields
+    std::string field_contents_prefix_; // For mapping to control fields
+    bool required_;
+public:
+    explicit FieldDescriptor(const std::string &name, const std::string &tag, const std::string &overflow_tag, const char indicator1,
+                             const char indicator2, const bool repeat_field,
+                             const std::vector<std::pair<char, std::string>> &subfield_codes_to_json_tags,
+                             const std::vector<std::pair<char, std::string>> &subfield_codes_to_prefixes,
+                             const std::vector<std::pair<char, std::string>> &subfield_codes_to_fixed_subfields,
+                             const std::string &json_tag, const std::string &field_contents_prefix, const bool required);
+    bool operator<(const FieldDescriptor &other) const { return tag_ < other.tag_; }
+};
+
+
+FieldDescriptor::FieldDescriptor(const std::string &name, const std::string &tag, const std::string &overflow_tag, const char indicator1,
+                                 const char indicator2, const bool repeat_field,
+                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_json_tags,
+                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_prefixes,
+                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_fixed_subfields,
+                                 const std::string &json_tag, const std::string &field_contents_prefix, const bool required)
+    : name_(name), tag_(tag), overflow_tag_(overflow_tag), indicator1_(indicator1), indicator2_(indicator2), repeat_field_(repeat_field),
+      subfield_codes_to_json_tags_(subfield_codes_to_json_tags), subfield_codes_to_prefixes_(subfield_codes_to_prefixes),
+      subfield_codes_to_fixed_subfields_(subfield_codes_to_fixed_subfields), json_tag_(json_tag),
+      field_contents_prefix_(field_contents_prefix), required_(required)
+{
+    if (not overflow_tag_.empty() and repeat_field_)
+        LOG_ERROR("field \"" + name_ + "\" can't have both, an over flow tag and being a repeat field!");
+
+    if (subfield_codes_to_json_tags_.empty() and json_tag_.empty())
+        LOG_ERROR("field \"" + name_
+                  + "\" neither a mapping to the contents of a control field nor the contents of data subfields has been specified!");
```

Comment:
"... nor have the contents of data subfields specified" would be more correct.

Reply:
Please have a look at the new wording...